### ISSUE TITLE
ManageIQ causes too many sessions and causes lockout on iLO4

### ISSUE THREAD

**Issue by vwbusguy** [ManageIQ/manageiq-providers-redfish#113]
On iLO4 (2.76, ProLiant BL460c Gen9), after a successful authentication, when manageIQ begins to refresh inventory, it triggers several new sessions and keeps them active, causing iLO to reach the session limit and block any further logins to the API or web interface.

![image](https://user-images.githubusercontent.com/5474941/82715441-fda0b400-9c47-11ea-8c03-1267129d1bf1.png)

Is it possible for ManageIQ to re-use an existing session for an operation and then log out when the operation is complete?  This seems to be how other API services are interacting with iLO based on the event log.

**Comment by vwbusguy** [comment]
At first glance, it seems that it is re-authing within the nested loops here rather than authorizing once, getting a session, running the queries, and then logging out: 
https://github.com/ManageIQ/manageiq-providers-redfish/blob/master/spec/factories/ext_management_system.rb#L12

**Comment by vwbusguy** [comment]
For context:

> Failure to log out of iLO by either browsing to a different site or closing the browser also results in an idle connection. The iLO firmware supports a finite number of connections. Misuse of the Infinite timeout option might make iLO inaccessible to other users. Idle connections are recycled after they time out. This setting applies to local and directory users. Directory server timeout settings might preempt the iLO setting.

It doesn't appear that there's any way to change the max number of concurrent sessions.  At best, you can set to kill off the orphaned sessions after 15 minutes (default is 30).

https://support.hpe.com/hpesc/public/docDisplay?docLocale=en_US&docId=emr_na-a00045340en_us

**Comment by agrare** [comment]
Interesting, the redfish client should be cached for a refresh [ref](https://github.com/ManageIQ/manageiq-providers-redfish/blob/master/app/models/manageiq/providers/redfish/inventory/collector.rb#L6) so it only has one instance of the connection but maybe something weird is going on with the RedFishClient gem and each API call creating a new session?

**Comment by vwbusguy** [comment]
Interesting - This manageiq instance is currently running in a "development" environment, so I wonder if it's not caching the gem.

**Comment by tadeboro** [comment]
It was me who wrote the majority of this code. It has been a while since I last worked on this, but IIRC, we leak one Redfish session per refresh. The main reason for this is that I was unable to find a "hook" that would be called at the end of the refresh process so that we could terminate the session that we used to refresh the resources.

The rest of the code only interacts with the Redfish backend from the blocks supplied to `with_provider_connection`. That terminates the session after the block returns and prevents session leaks.

It would be possible to adapt the refresh process and make sure we do not leak sessions, but because of the Redfish API specifics, we would need to duplicate quite a lot of logic from the parser in the collector. In short, every subresource that parser touches would need to be fetched by the collector in advance. This might not seem like a big problem until one looks at how deeply nested the results from the API are.

To demonstrate how "bad" things are, let us look at the [get_server_disk_capacity](https://github.com/ManageIQ/manageiq-providers-redfish/blob/808a3f9326ebfdb321790e5e7c3b5ac34e453229/app/models/manageiq/providers/redfish/inventory/parser/physical_infra_manager.rb#L151)  method. This is the rough list of API calls that would be made on a Lenovo server that I have a Redfish recording of:

 1. 2 API calls to retrieve the storages (M.2 and RAID),
 2. 2 API calls to retrieve M.2 devices (there are 2 disks in this server),
 3. 16 API calls to retrieve disks that are under RAID controllers.

If we would like to avoid leaking a session, all these calls would need to be made in the collector, and the parser would then visit them in much the same fashion when mapping data to MIQ's internal data structures. That is 20 API calls and we only covered disks.

I had plans to fix this sometime in the future, but unfortunately, the real-world disagreed with my plans, and I was moved to work on other stuff.

I hope this information helps a bit. Good luck!

**Comment by tadeboro** [comment]
Just for reference: the commit that allowed us to logout after we are done is at https://github.com/ManageIQ/manageiq/commit/2619b1ea6e943fa4445a85c3584842c3b6e5ea49

And the code in the Redfish provider was added in https://github.com/ManageIQ/manageiq-providers-redfish/commit/2797a145dc1ae8b4a09983162a4fef1e9f492372

**Comment by tadeboro** [comment]
Why is Redfish provider having trouble fetching and refreshing the inventory? Because the OData specification that the Redfish API is based on is a true REST API. In such APIs, we can only fetch subresource once we parse the data from the parent resource. For example, to gather information about the CPUs in a server, we need to first parse a server resource since that resource contains links to CPU details. This means that parsing and retrieving information are relatively tightly coupled.

ManageIQ, on the other hand, assumes that collecting the resource data and parsing it are not coupled at all.

And I promise this is my last "I just remembered something else" comments ;)

**Comment by agrare** [comment]
Oh interesting @tadeboro I didn't realize that you didn't have a way to logout at the end of a refresh.

We have a `Collector#collect` method that can be used to pre-collect all inventory for parsing, if it is feasible we could do the collection there then logout at the end of the method.

The other alternative is to expose the connection as a public attribute of the Collector class then have the parser logout at the end of the `Parser#parse` method.  This would be the simplest I think.

**Comment by agrare** [comment]
@vwbusguy I don't want to say that https://github.com/ManageIQ/manageiq-providers-redfish/pull/114 fixes this issue because there could be other issues but if you get a chance could you test with that change applied and see if it resolves your issues?  Thanks!

**Comment by vwbusguy** [comment]
@agrare @tadeboro - Thank you both for the very quick turnaround in looking at this!  I'll try to test it out this afternoon and let you know.

**Comment by vwbusguy** [comment]
I clone the repo at 5d3a86c8bfae2837ad4312d973a03d87a93858e4 and built it.  I set the gem override and restarted manage iq.  (I saw the messages about the gem override at on restart.)  I connected to a different ESX iLO4 host and authentication initially worked, but then the same thing happened.  These appear to be hard-coded at 15 concurrent sessions.  I had logged in directly to the iLO interface and almost instantly after I connected it to ManageIQ, the other sessions were taken up.  So, it doesn't appear to be logging out with this change and the initial sync process ends up dying part-way through.

If it's helpful, I don't know if I mentioned this above, but this manageiq instance is currently running ivanchuk-5 (I saw a jansa/yes? tag in the PR).

**Comment by agrare** [comment]
If you saw it instantly fill up then it isn't likely because of refresh leaking a single session every time.  I think it is every 15 minutes so one session per refresh would take almost 4 hours.

Could you try not running the whole server but individual operations so that we can try to diagnose this?

```
# systemctl stop evmserverd # This will shut the server down so workers aren't running things in the background
# cd /var/www/miq/vmdb
# bundle exec rails c
>> redfish = ManageIQ::Providers::Redfish::PhysicalInfraManager.first
>> redfish.with_provider_connection {} # This will simulate a simple login/logout
>> redfish.authentication_check        # This will simulate doing an auth check
>> EmsRefresh.refresh(redfish)         # This will run a single full refresh
```

**Comment by vwbusguy** [comment]
`.with_provider_connection {}` returned nil
`.authentication_check` returned `=> [true, ""]`
Both of these actions spawned a login to iLO that persisted after the action was complete.  (There are two logins still in "Active Sessions" in the iLO web interface from the ManageIQ server.)

Unfortunately, it seems I'm missing a step for the last bit and I'm not familiar enough with the code based to recognize what I'm missing:
```ruby
irb(main):004:0> EmsRefresh.refresh(redfish)  
Traceback (most recent call last):
       16: from bin/rails:4:in `require'
       15: from railties (5.1.7) lib/rails/commands.rb:16:in `<top (required)>'
       14: from railties (5.1.7) lib/rails/command.rb:44:in `invoke'
       13: from railties (5.1.7) lib/rails/command/base.rb:63:in `perform'
       12: from thor (1.0.1) lib/thor.rb:392:in `dispatch'
       11: from thor (1.0.1) lib/thor/invocation.rb:127:in `invoke_command'
       10: from thor (1.0.1) lib/thor/command.rb:27:in `run'
        9: from railties (5.1.7) lib/rails/commands/console/console_command.rb:97:in `perform'
        8: from railties (5.1.7) lib/rails/commands/console/console_command.rb:17:in `start'
        7: from railties (5.1.7) lib/rails/commands/console/console_command.rb:62:in `start'
        6: from (irb):4
        5: from app/models/ems_refresh.rb:102:in `refresh'
        4: from app/models/ems_refresh.rb:102:in `each'
        3: from app/models/ems_refresh.rb:103:in `block in refresh'
        2: from app/models/manageiq/providers/base_manager/refresher.rb:11:in `refresh'
        1: from app/models/manageiq/providers/base_manager/refresher.rb:67:in `refresh'
ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError (uninitialized constant ManageIQ::Providers::Redfish::PhysicalInfraManager::RefreshParser)
```

**Comment by vwbusguy** [comment]
Here's the full log:
```
# bin/bundle exec rails c
** override_gem("manageiq-providers-redfish", :path=>"/opt/manageiq/manageiq-providers-redfish") at /opt/manageiq/manageiq/bundler.d/overrides.rb:1
** override_gem("manageiq-providers-redfish", :path=>"/opt/manageiq/manageiq-providers-redfish") at /opt/manageiq/manageiq/bundler.d/overrides.rb:1
** Using session_store: ActionDispatch::Session::MemCacheStore
** ManageIQ ivanchuk-5, codename: Ivanchuk
Loading development environment (Rails 5.1.7)
irb(main):001:0> redfish = ManageIQ::Providers::Redfish::PhysicalInfraManager.first
PostgreSQLAdapter#log_after_checkout, connection_pool: size: 5, connections: 1, in use: 1, waiting_in_queue: 0
=> #<ManageIQ::Providers::Redfish::PhysicalInfraManager id: 673, name: "esx-(redacted)", created_on: "2020-05-22 22:14:48", updated_on: "2020-05-26 19:54:15", guid: "622a61dd-28a9-4732-90c5-e91b085cc0c0", zone_id: 2, type: "ManageIQ::Providers::Redfish::PhysicalInfraManager", api_version: nil, uid_ems: nil, host_default_vnc_port_start: nil, host_default_vnc_port_end: nil, provider_region: nil, last_refresh_error: "undefined method `FirmwareInventory' for nil:NilCl...", last_refresh_date: "2020-05-26 19:54:15", provider_id: nil, realm: nil, tenant_id: 1, project: nil, parent_ems_id: nil, subscription: nil, last_metrics_error: nil, last_metrics_update_date: nil, last_metrics_success_date: nil, tenant_mapping_enabled: nil, enabled: true, options: {}, zone_before_pause_id: nil, last_inventory_date: nil>
irb(main):002:0> redfish.with_provider_connection {}
=> nil
irb(main):003:0> redfish.authentication_check
/usr/share/ruby/securerandom.rb:273: warning: already initialized constant Random::Formatter::ALPHANUMERIC
/usr/share/ruby/securerandom.rb:273: warning: previous definition of ALPHANUMERIC was here
=> [true, ""]
irb(main):004:0> EmsRefresh.refresh(redfish)  
Traceback (most recent call last):
       16: from bin/rails:4:in `require'
       15: from railties (5.1.7) lib/rails/commands.rb:16:in `<top (required)>'
       14: from railties (5.1.7) lib/rails/command.rb:44:in `invoke'
       13: from railties (5.1.7) lib/rails/command/base.rb:63:in `perform'
       12: from thor (1.0.1) lib/thor.rb:392:in `dispatch'
       11: from thor (1.0.1) lib/thor/invocation.rb:127:in `invoke_command'
       10: from thor (1.0.1) lib/thor/command.rb:27:in `run'
        9: from railties (5.1.7) lib/rails/commands/console/console_command.rb:97:in `perform'
        8: from railties (5.1.7) lib/rails/commands/console/console_command.rb:17:in `start'
        7: from railties (5.1.7) lib/rails/commands/console/console_command.rb:62:in `start'
        6: from (irb):4
        5: from app/models/ems_refresh.rb:102:in `refresh'
        4: from app/models/ems_refresh.rb:102:in `each'
        3: from app/models/ems_refresh.rb:103:in `block in refresh'
        2: from app/models/manageiq/providers/base_manager/refresher.rb:11:in `refresh'
        1: from app/models/manageiq/providers/base_manager/refresher.rb:67:in `refresh'
ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError (uninitialized constant ManageIQ::Providers::Redfish::PhysicalInfraManager::RefreshParser)
```

**Comment by agrare** [comment]
Oh that error is because you're running a master plugin but an ivanchuk core, you can try the test with the old plugin.

**Comment by vwbusguy** [comment]
Ah, OK - I'll remove the gem override and try again in a few minutes.

**Comment by agrare** [comment]
> Both of these actions spawned a login to iLO that persisted after the action was complete.

This sounds like an issue, because `.with_provider_connection` calls logout after.  It sounds like the logout action isn't working.

You can also try:
```
>> redfish.connect.logout
```

**Comment by vwbusguy** [comment]
So reverting back to ivanchuk-5 for redfish, I got a different error on refresh:
```ruby
irb(main):005:0> EmsRefresh.refresh(redfish) 
Traceback (most recent call last):
       16: from bin/rails:4:in `require'
       15: from railties (5.1.7) lib/rails/commands.rb:16:in `<top (required)>'
       14: from railties (5.1.7) lib/rails/command.rb:44:in `invoke'
       13: from railties (5.1.7) lib/rails/command/base.rb:63:in `perform'
       12: from thor (1.0.1) lib/thor.rb:392:in `dispatch'
       11: from thor (1.0.1) lib/thor/invocation.rb:127:in `invoke_command'
       10: from thor (1.0.1) lib/thor/command.rb:27:in `run'
        9: from railties (5.1.7) lib/rails/commands/console/console_command.rb:97:in `perform'
        8: from railties (5.1.7) lib/rails/commands/console/console_command.rb:17:in `start'
        7: from railties (5.1.7) lib/rails/commands/console/console_command.rb:62:in `start'
        6: from (irb):5
        5: from app/models/ems_refresh.rb:102:in `refresh'
        4: from app/models/ems_refresh.rb:102:in `each'
        3: from app/models/ems_refresh.rb:103:in `block in refresh'
        2: from app/models/manageiq/providers/base_manager/refresher.rb:11:in `refresh'
        1: from app/models/manageiq/providers/base_manager/refresher.rb:67:in `refresh'
ManageIQ::Providers::BaseManager::Refresher::PartialRefreshError (undefined method `FirmwareInventory' for nil:NilClass)
```

However, the more interesting thing is that each time I ran the logout:
```ruby
irb(main):007:0> redfish.connect.logout
=> ["Authorization", "X-Auth-Token"]
```
...on the iLO4 side, rather than logging out the existing session, it instead generated a new active session that also stuck around.

**Comment by vwbusguy** [comment]
Seems it could be related to this: https://github.com/xlab-si/redfish-client-ruby/issues/32

**Comment by agrare** [comment]
:+1: yeah it sounds like it

**Comment by vwbusguy** [comment]
OK, so here's what I think I've found out.  The author of the ruby client and HPE has a "disagreement" over the RedFish spec and seem to have no interest in meeting each other partway.  HPE is passing the Session in location where the ruby spec expects it in the body - the spec seems to allow for either.  The ruby plugin author seems to expect that HPE will put it in the body and HPE seems to expect that any client will use Location.

Here's the ref from HPE: https://developer.hpe.com/blog/managing-ilo-sessions-with-redfish (Note the passing swipe):

> At some point of time, if you call again and again such a badly written script, you will reach the iLO maximum number of sessions and get completely locked out until the iLO is physically reset or until sessions are removed when the UserExpires time is reached.

Note that this isn't the only affected manufacturer as there is also an incompatibility with IBM servers over a response code mismatch 200/201.  The plugin author seems resistant to incorporate any changes he sees as "vendor specific".

There are other Ruby libraries, including an iLO one that supports iLO4 (but not iLO5 natively) that can also talk RedFish, but hasn't abstracted the RedFish calls (https://github.com/HewlettPackard/ilo-sdk-ruby) or the current ruby client could possible be forked and with maybe 3 LOC changes (I haven't looked closer) support RedFish implementations from IBM and HPE.  

Ideally, we would be able to work with the ruby client author to better it for everyone, but the iLO issue was submitted over a year ago and seems unlikely to be picked up.  Otherwise, I could also just patch it locally to work with iLO4 here, but I don't have the bandwidth or Ruby background to maintain a proper upstream library.

**Comment by agrare** [comment]
Badly written is certainly relative since from my reading of the spec it sounds like both sides aren't entirely correct.

I commented in the iLO issue on the redfish-client so hopefully we can get some traction there.

**Comment by tadeboro** [comment]
I am the "maintainer" of the Redfish client library in question. I put maintainer in quotes because I originally wrote it and then got stuck maintaining it even if I do not use it anymore for my day-to-day work. And yes, I am highly opinionated when it comes to accepting vendor-specific hacks because I do need to maintain each and every drive-by fix that people are so keen on contributing.

But you are being a bit dishonest here. The [iLO4 problem](https://github.com/xlab-si/redfish-client-ruby/issues/32) is not on the "I do not want to fix it" list. I clearly stated that I did not have access to the iLO4 and lacked the time to do it myself (and now that I do not work on MIQ anymore, this time is even more limited). I also offered to review the contributed changes but there were no contributions. I even started working on [a fix](https://github.com/xlab-si/redfish-client-ruby/commit/0e08163ee19e1c23ee373391f16652f6551d9279) some time ago, but never got it tested with the real system, so it never got merged to master.

The [IBM issue](https://github.com/xlab-si/redfish-client-ruby/issues/44) you are referencing is a bit different because returning 200 when a backend creates a resource is just plain wrong. The 201 status exists solely for this purpose. And even then, I offered to accept the "fix" under some reasonable conditions. Asking to file an upstream bug before adding a hack should not be such a big deal.

So, in short:

 1. I will review changes that would not go against the standard.
 2. I will not add a hack to a library without proper justification. Reporting a bug to vendor is a must here.
 3. Yes, it is quite painful to work with me most of the time ;)

**Comment by agrare** [comment]
I also don't have access to an iLO to test with, @vwbusguy could you try https://github.com/xlab-si/redfish-client-ruby/pull/46 ?

**Comment by vwbusguy** [comment]
@tadeboro - Thank you for your time in looking at this and clarifying.  I apologize for misrepresenting the situation and I appreciate the work you have done here.  I know I'm not in a place to maintain something like this myself either, so I respect that you have continued to do so even for something you aren't actively using.

@agrare - Thank you for doing this - I would be happy to test out your patch against one of our iLO4 environments.

**Comment by vwbusguy** [comment]
Many thanks to @agrare for the upstream PR.  It looks like there's a fix in https://github.com/xlab-si/redfish-client-ruby/releases/tag/0.5.2 - I'll see if I can override that dependency and test it out in manageiq this afternoon.

**Comment by tadeboro** [comment]
Our CI/CD process also published the 0.5.2 version to https://rubygems.org/gems/redfish_client/versions/0.5.2, and MIQ should pick that version up on the next nightly build.

**Comment by vwbusguy** [comment]
Awesome - many thanks, @tadeboro !

**Comment by vwbusguy** [comment]
I have good news in that the session issue appears to be fixed (ivanchuk-5 tag on this repo with the redfish_client bumped to 0.5.2 in the spec).  I'm only seeing 1-2 ManageIQ sessions at any given time per iLO instance connected.

However, it seems that the refreshes are failing due to:
`[----] E, [2020-05-27T14:34:21.194646 #3218990:3f80a1a5c1f0] ERROR -- : MIQ(ManageIQ::Providers::Redfish::PhysicalInfraManager::EventCatcher::Runner#start_event_monitor) EMS [x.x.x.x] as [Administrator] Event Monitor Thread aborted because [undefined method `listen' for nil:NilClass]`

I think this error might be unrelated.

**Comment by vwbusguy** [comment]
Ah, yes, I'm now seeing the issue reported in #90 .  This one seems fixed.  Many thanks @tadeboro and @agrare !

**Comment by agrare** [comment]
:tada: awesome news thanks for helping us test it out @vwbusguy
